[release/7.0] Add and use RDF.InferMetadata and update new PopulateMetadata API

halter73 opened the pull request:
Fixes #43000 and #43125

halter73 committed:
Expose entire EndpointBuilder to filter factories

halter73 committed:
Apply suggestions from code review

Co-authored-by: Brennan <brecon@microsoft.com>

halter73 committed:
Remove redundant if/else

halter73 committed:
Fix ArgumentException

halter73 committed:
Add and use RDF.InferMetadata

- Stop flowing EndpointBuilder to filter factories

halter73 committed:
Update PopulateMetadata API

halter73 commented on src/Http/Http.Abstractions/src/PublicAPI.Unshipped.txt:
```diff
@@ -5,7 +5,7 @@ abstract Microsoft.AspNetCore.Http.EndpointFilterInvocationContext.Arguments.get
 abstract Microsoft.AspNetCore.Http.EndpointFilterInvocationContext.HttpContext.get -> Microsoft.AspNetCore.Http.HttpContext!
 Microsoft.AspNetCore.Builder.EndpointBuilder.ApplicationServices.get -> System.IServiceProvider!
-Microsoft.AspNetCore.Builder.EndpointBuilder.ApplicationServices.set -> void
+Microsoft.AspNetCore.Builder.EndpointBuilder.ApplicationServices.init -> void
```
This isn't approved yet: https://github.com/dotnet/aspnetcore/issues/43544

It's trivial to change this back to set if we reject the API, but I think `init` makes more sense.

halter73 replied:
This has now been approved.

halter73 commented on src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:
```diff
@@ -6520,48 +6649,79 @@ public static void PopulateMetadata(EndpointMetadataContext context)
+    // bool TryParse(string, IFormatProvider, out AddsCustomParameterMetadata)
+    // bool TryParse(string, out AddsCustomParameterMetadata)
+    // but found
+    // static Boolean TryParse(System.String, System.IFormatProvider, AddsCustomParameterMetadata ByRef)
```
I filed an issue for this: #43546

@dotnet/minimal-apis

brunolins16 commented on src/Shared/EndpointMetadataProviderRunner.cs:
```diff
@@ -0,0 +1,72 @@
+
+namespace Microsoft.AspNetCore.Http;
+
+internal static class EndpointMetadataPopulator
```
Is this a dup? https://github.com/dotnet/aspnetcore/blob/1f057843d64b59c9dfe61ce4892ca9afaa664428/src/Shared/EndpointMetadataPopulator.cs

halter73 replied:
Good catch!

halter73 committed:
Delete redundant file

halter73 committed:
Suppress trimmer warnings

- We do the same ParameterBindingMethodCache
- We were calling MakeGenericMethod the same way before in RDF and
  EndpointMetadataConvention. The difference is that it's now shared source.

halter73 committed:
Add tests back

BrennanConroy commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -158,9 +194,28 @@ public static RequestDelegateResult Create(Delegate handler, RequestDelegateFact
+    /// The result returned from <see cref="InferMetadata(MethodInfo, RequestDelegateFactoryOptions?)"/> if that was used to inferring metadata before creating the final RequestDelegate.
+    /// If <see langword="null"/>, this call to <see cref="Create(Delegate, RequestDelegateFactoryOptions?, RequestDelegateMetadataResult?)"/> method will infer the metadata that
+    /// <see cref="InferMetadata(MethodInfo, RequestDelegateFactoryOptions?)"/> would have inferred for the same <see cref="Delegate.Method"/> and populate <see cref="RequestDelegateFactoryOptions.EndpointBuilder"/>
+    /// with that metadata. Otherwise, this metadata inference be skipped as this step has already been done.
```
```suggestion
    /// with that metadata. Otherwise, this metadata inference will be skipped as this step has already been done.
```

BrennanConroy commented on src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:
```diff
@@ -6228,27 +6237,117 @@ public void Create_CombinesAllMetadata_InCorrectOrder()
+    }
+
+    [Fact]
+    public void InferMetadata_ThenCreate_CombinesAllMetadata_InCorrectOrder()
```
I could be missing it, but is there a test for skipping when metadata has already been inferred?

halter73 replied:
Create_DoesNotInferMetadata_GivenManuallyConstructedMetadataResult is that, and a bunch of tests using RouteEndpointDataSource verify this path indirectly since that's what always happens in the default case now.

BrennanConroy commented on src/Http/Routing/src/RouteEndpointBuilder.cs:
```diff
@@ -30,10 +30,12 @@ public sealed class RouteEndpointBuilder : EndpointBuilder
     /// <param name="order">The order assigned to the endpoint.</param>
     public RouteEndpointBuilder(
-       RequestDelegate requestDelegate,
+       RequestDelegate? requestDelegate,
```
I think this was primarily done because we pass a potentially null `RequestDelegate` already:
https://github.com/dotnet/aspnetcore/blob/318e16b6ac3a9644b266a463c8db47d6a24cad0d/src/Http/Routing/src/EndpointDataSource.cs#L57
EndpointDataSource should probably be updated to not use `!` now.

halter73 replied:
When we approved this in #43000, we discussed leaving it annotated as non-nullable, but we felt uncomfortable actually throwing an ArgumentNullException if it was null after not doing so for so long. The EndpointDataSource code shows how that kind of thing could break working code. And since `RouteEndpointBuilder.RequestDelegate` is nullable anyway, there's really no downside to just fixing the nullability annotation here in the RouteEndpointBuilder constructor.

Good catch about the unneeded `!` in `EndpointDataSource.GetGroupedEndpoints`. I'll remove it.

Pilchie commented:
@halter73 - this was approved for RC2.

halter73 commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -279,11 +355,7 @@ private static IReadOnlyList<object> AsReadOnlyList(IList<object> metadata)
-            {
-                return null;
-            }
+            return null;
```
Since we [stopped passing normal RequestDelegates to RDF just because it returned a `Task<T>`](https://github.com/dotnet/aspnetcore/pull/42519/files#diff-f8807c470bcc3a077fb176668a46df57b4bb99c992b6b7b375665f8bf3903c94L202-L207), we no longer have to check for that case here and can preserve the original RequestDelegate any time the filter pipeline returns the original `next` invocation.

halter73 commented on src/Http/Http.Abstractions/src/EndpointFilterFactoryContext.cs:
```diff
@@ -13,34 +12,21 @@ namespace Microsoft.AspNetCore.Http;
-    public IServiceProvider ApplicationServices { get; }
+    public IServiceProvider ApplicationServices { get; init; } = EmptyServiceProvider.Instance;
+
+    private sealed class EmptyServiceProvider : IServiceProvider
```
But they're so small! And shared source doesn't unbload the assembly anyway.

halter73 committed:
Apply suggestions from code review

Co-authored-by: Brennan <brecon@microsoft.com>

halter73 committed:
Address more PR feedback

BrennanConroy commented:
/azp run

halter73 commented:
I'm merging this using my build-ops permissions since the PR has been approved by tactics, the PR has been approved by code reviewers, and the API has been approved by API reviewers.
